### DISCUSSION TITLE
Add automatically completion as language feature.

### DISCUSSION THREAD

**Discussion by YagaoDirac** [carbon-language/carbon-lang#1791] [Language design]
I saw some discussion about reflection and also some new ideas on new ways to make classes.
All these ideas are great. The surprising maybe controllable. But the issue I see is that, most of these ideas mess the auto completion.
I'm a super big fan of auto completion. I believe it's the top tool. It has to be part of a language.
My idea is that, if the auto completion is part of the language, then it's easy to control it with the same syntax. If any design messes up with the auto completion, we can manually fix it with some code. The IDE is only in charge of representing some gui for auto completion.
Also, this forces the language team to design the language along with auto completion which is great. This also ensure the regularity. Then, regularity reduces guess work.

**Comment by jonmeow** [comment]
This is already a goal. However, autocompletion is more of an editor feature than a language feature. In the project goals, you'll find relevant mentions under developer tooling as part of the ecosystem: https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/goals.md#language-tools-and-ecosystem

**Reply by YagaoDirac** [reply]
You still plan to leave the auto completion mainly to the IDE developer, this is why I have to provide my opinion here. IDE developers worked also very hard for many many years but auto completion systems work almost only for pure original version of languages. As I know, in UE4 project, wxWidget project, or anything that has a lot macros, auto completion rarely works.
The only way I can imagine is it's part of the language. When people make libraries, they also make auto completion instruction along with the libraries.
Also, I wrote a suggestion to vs2022 asked for small icons for the keywords to enhance readability. I think the readability is so hard to achieve with only text design. This is probably the way to utilize IDEs.
